accepted/00000-ptf-repos-in-uyuni-2.md

```diff
@@ -0,0 +1,179 @@
+
+When we provide these PTFs in SUSE Manager as channels, it is easier to roll out the fixes to multiple clients.
+We need to take care that the PTFs can be installed, updated and been replaced by a regular maintenance update which contains the fix.
+
+Sometimes SUSE provide also TEST packages. The difference is, that they are just meant for testing on a single system to better identify the problem and gather logfiles.
+These packages are not supported and should not rolled out on multiple systems.
+It is planned to have them in separate channels. We need to provide a best practice to work with them.
+We need to be able to install and update them. Replacing them by a PTF or a maintenance update needs to be possible.
+
+# Detailed design
+[design]: #detailed-design
+
+## General PTF design for SUSE OSes
+
+A ptf release consists of multiple rpm packages plus one
+master "ptf" package. The master ptf package has two
+purposes:
+
+- it allows to easily query which ptfs are installed
+  in the system
+- it makes sure that all of the installed rpm packages
+  come from the same release, i.e. there is no mix up
+  between multiple releases are non-ptf package
+
+To do this the master ptf package contains the following
+elements (incident 1234, first release):
+
+    Name: ptf-1234
+    Version: 1
+    Release: 1                 # always 1
+    Provides: ptf() = 1234-1
+    Requires: (pkg1 = pkg1EVR if pkg1)
+    Requires: (pkg2 = pkg1EVR if pkg2)
+    ...
+
+I.e. if pkg1 is installed it must be installed with version
+pkg1EVR. We do not use `Conflict: pkg1 != pkg1EVR` for SUSE because that
+would not work with multiversion packages like the kernel.
+But for Non-SUSE systems and packages, which do not support multiversion
+packages, we could generate `Conflicts` dependencies.
+
+All packages providing ptf() get blacklisted by the solver,
+meaning they can only be installed by a specific solver job
+that addresses them. This means that selecting a specific
+ptf master package via yast or "zypper in" works, but they
+can not be pulled in via dependencies.
+
+Each individual rpm package must require the specific
+master ptf package:
+
+    Name: pkg1
+    Requires: ptf-1234 = 1-1
+
+This makes sure that the solver cannot pull in the rpm
+packages, as that would mean to also pull in the blacklisted
+master ptf package.
+
+
+
+### Updates
+
+Ptfs can be updated to a new release by calling 'zypper up'.
+This will update the master ptf package to some higher
+version (e.g. ptf-1234 = 2-1) and also pull in the corresponding
+rpm packages.
+
+This should work with any installer.
+
+### Uninstalling ptfs
+
+'zypper rm ptf-1234' will uninstall the ptf and revert back to
+non-ptf packages.
+
+We need to check if this is working out of the box or if we need to
+implement something in Uyuni to make it work.
```

Comment:
Could this in principle be just a documented best practice for starters, in worst case?